# Atividade 3: Verificar se é possível melhorar os resultados de aprendizado fazendo transformações em atributos numéricos

* **Nome**: Matheus Freitas Martins
* **Matrícula**: ES111281

## 1. Ler os dados do arquivo “churn-telecom.csv” e fazer uma análise exploratória inicial, para conhecer os atributos

In [240]:
# Importando bibliotecas necessárias
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

# Carregar os dados
df = pd.read_csv('churn-telecom.csv')

In [241]:
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [242]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

In [243]:
df.isnull().sum()

state                     0
account length            0
area code                 0
phone number              0
international plan        0
voice mail plan           0
number vmail messages     0
total day minutes         0
total day calls           0
total day charge          0
total eve minutes         0
total eve calls           0
total eve charge          0
total night minutes       0
total night calls         0
total night charge        0
total intl minutes        0
total intl calls          0
total intl charge         0
customer service calls    0
churn                     0
dtype: int64

In [244]:
df.value_counts()

state  account length  area code  phone number  international plan  voice mail plan  number vmail messages  total day minutes  total day calls  total day charge  total eve minutes  total eve calls  total eve charge  total night minutes  total night calls  total night charge  total intl minutes  total intl calls  total intl charge  customer service calls  churn
AK     1               408        373-1028      no                  no               0                      175.2              74               29.78             151.7              79               12.89             230.5                109                10.37               5.3                 3                 1.43               1                       False    1
NV     87              510        331-8484      no                  yes              39                     82.6               113              14.04             224.4              63               19.07             163.6                88                 7.36         

In [245]:
df.describe()

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [246]:
for column in df.columns:
    unique_values = df[column].unique()
    print(f"Valores únicos na coluna '{column}': {unique_values}")

Valores únicos na coluna 'state': ['KS' 'OH' 'NJ' 'OK' 'AL' 'MA' 'MO' 'LA' 'WV' 'IN' 'RI' 'IA' 'MT' 'NY'
 'ID' 'VT' 'VA' 'TX' 'FL' 'CO' 'AZ' 'SC' 'NE' 'WY' 'HI' 'IL' 'NH' 'GA'
 'AK' 'MD' 'AR' 'WI' 'OR' 'MI' 'DE' 'UT' 'CA' 'MN' 'SD' 'NC' 'WA' 'NM'
 'NV' 'DC' 'KY' 'ME' 'MS' 'TN' 'PA' 'CT' 'ND']
Valores únicos na coluna 'account length': [128 107 137  84  75 118 121 147 117 141  65  74 168  95  62 161  85  93
  76  73  77 130 111 132 174  57  54  20  49 142 172  12  72  36  78 136
 149  98 135  34 160  64  59 119  97  52  60  10  96  87  81  68 125 116
  38  40  43 113 126 150 138 162  90  50  82 144  46  70  55 106  94 155
  80 104  99 120 108 122 157 103  63 112  41 193  61  92 131 163  91 127
 110 140  83 145  56 151 139   6 115 146 185 148  32  25 179  67  19 170
 164  51 208  53 105  66  86  35  88 123  45 100 215  22  33 114  24 101
 143  48  71 167  89 199 166 158 196 209  16  39 173 129  44  79  31 124
  37 159 194 154  21 133 224  58  11 109 102 165  18  30 176  47 190 152
  26  

* **state**: Atributo categórico que representa o estado dos EUA em que o cliente reside.
* **account length**: Atributo numérico que indica a duração da conta do cliente em meses.
* **area code**: Atributo numérico que indica o código de área do telefone do cliente.
* **phone number**: Atributo categórico que representa o número de telefone do cliente.
* **international plan**: Atributo categórico que indica se o cliente possui um plano internacional (yes/no).
* **voice mail plan**: Atributo categórico que indica se o cliente possui um plano de correio de voz (yes/no).
* **number vmail messages**: Atributo numérico que representa o número de mensagens de correio de voz que o cliente possui.
* **total day minutes**: Atributo numérico que representa a quantidade total de minutos que o cliente usa durante o dia.
* **total day calls**: Atributo numérico que representa a quantidade total de chamadas feitas pelo cliente durante o dia.
* **total day charge**: Atributo numérico que representa a cobrança total do cliente pelas chamadas feitas durante o dia.
* **total eve minutes**: Atributo numérico que representa a quantidade total de minutos que o cliente usa durante a noite.
* **total eve calls**: Atributo numérico que representa a quantidade total de chamadas feitas pelo cliente durante a noite.
* **total eve charge**: Atributo numérico que representa a cobrança total do cliente pelas chamadas feitas durante a noite.
* **total night minutes**: Atributo numérico que representa a quantidade total de minutos que o cliente usa durante a noite.
* **total night calls**: Atributo numérico que representa a quantidade total de chamadas feitas pelo cliente durante a noite.
* **total night charge**: Atributo numérico que representa a cobrança total do cliente pelas chamadas feitas durante a noite.
* **total intl minutes**: Atributo numérico que representa a quantidade total de minutos que o cliente usa para chamadas internacionais.
* **total intl calls**: Atributo numérico que representa a quantidade total de chamadas internacionais feitas pelo cliente.
* **total intl charge**: Atributo numérico que representa a cobrança total do cliente pelas chamadas internacionais feitas.
* **customer service calls**: Atributo numérico que representa o número de chamadas de atendimento ao cliente feitas pelo cliente.
* **churn**: Atributo categórico que indica se o cliente cancelou o serviço (True/False).

In [247]:
# Função para encapsular o processo de treinamento e teste do classificador
def train_and_evaluate(X, y):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) # Dividindo os dados em conjuntos de treinamento e teste, usando a função train_test_split
    clf = RandomForestClassifier(random_state=42) # Criando o classificador Random Forest
    clf.fit(X_train, y_train) # Ajustando o classificador aos dados de treinamento.
    y_pred = clf.predict(X_test) # Fazendo previsões no conjunto de teste.
    accuracy = accuracy_score(y_test, y_pred) 
    return accuracy

# Função com Scaled para encapsular o processo de treinamento e teste do classificador
def train_and_evaluate_scaled(X, y):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    clf = RandomForestClassifier(random_state=42)
    clf.fit(X_train_scaled, y_train)
    y_pred = clf.predict(X_test_scaled)
    accuracy = accuracy_score(y_test, y_pred)
    return accuracy

### Pré-processamento

In [248]:
# Convertendo as colunas 'international plan' e 'voice mail plan' em valores numéricos (1 para 'yes' e 0 para 'no').
df['international plan'] = df['international plan'].map({'yes': 1, 'no': 0})
df['voice mail plan'] = df['voice mail plan'].map({'yes': 1, 'no': 0})

# Convertendo a coluna 'churn' em um tipo de dados inteiro.
df['churn'] = df['churn'].astype(int)

## 2. Aplicar um algoritmo de classificação de sua escolha e calcular a precisão. A variável alvo para ser classificada é “churn”
"Churn" é um termo utilizado em negócios para descrever a taxa de clientes que param de usar um produto ou serviço durante um determinado período de tempo.

Algoritmo de classificação escolhido: Random Forest

### Dividindo os dados em conjuntos de treinamento e teste

In [249]:
# Selecionando as colunas necessárias para treinar o modelo (X) e a variável de destino (y).
X = df.drop(['state', 'area code', 'phone number', 'churn'], axis=1)
y = df['churn']

In [250]:
# Calculando a acurácia antes das transformações
accuracy_before = train_and_evaluate_scaled(X, y)
print(f"Acurácia antes das transformações: {accuracy_before}")

Acurácia antes das transformações: 0.9475262368815592


## 3. Realizar transformações (arredondamento, tornar binário, binning) em atributos numéricos que aparentemente podem se beneficiar dessas transformações

In [251]:
# Realizando transformações nos atributos numéricos
# Arredondando os valores das colunas 'total day minutes', 'total eve minutes', 'total night minutes' e 'total intl minutes' para números inteiros
df['total day minutes'] = df['total day minutes'].apply(lambda x: round(x))
df['total eve minutes'] = df['total eve minutes'].apply(lambda x: round(x))
df['total night minutes'] = df['total night minutes'].apply(lambda x: round(x))
df['total intl minutes'] = df['total intl minutes'].apply(lambda x: round(x))

# convertendo a coluna "number vmail messages" em uma coluna binária, onde os valores serão 1 se houver mensagens de correio de voz e 0 se não houver mensagens de correio de voz.
df['number vmail messages'] = df['number vmail messages'].apply(lambda x: 1 if x > 0 else 0)

## 4. Aplicar novamente o algoritmo de classificação e comparar os resultados

In [252]:
# Dividindo os dados em conjuntos de treinamento e teste novamente após as transformações
X = df.drop(['state', 'area code', 'phone number', 'churn'], axis=1)

# Calculando a acurácia após as transformações
accuracy_after = train_and_evaluate_scaled(X, y)
print(f"Acurácia após transformações: {accuracy_after}")

# Comparando os resultados
print(f"Diferença na acurácia: {accuracy_after - accuracy_before}")

Acurácia após transformações: 0.9550224887556222
Diferença na acurácia: 0.007496251874062998


## 5. Repetir os passos 3 e 4 para diferentes atributos e formas de transformação

In [253]:
# Definindo o número de intervalos
num_intervals = 3
labels = ['baixo', 'médio', 'alto']
# Realizando o binning usando o método qcut, dividindo os dados em intervalos de quantis igualmente espaçados.
df['total day minutes'] = pd.qcut(df['total day minutes'], num_intervals, labels=labels)
# Convertendo a coluna 'total day minutes category' para códigos numéricos
df['total day minutes'] = df['total day minutes'].astype('category').cat.codes

# Realizando o binning usando o método qcut diretamente na coluna 'total eve minutes'
df['total eve minutes'] = pd.qcut(df['total eve minutes'], num_intervals, labels=labels)
# Convertendo a coluna 'total eve minutes' para códigos numéricos
df['total eve minutes'] = df['total eve minutes'].astype('category').cat.codes

# Realizando o binning usando o método qcut diretamente na coluna 'total night minutes'
df['total night minutes'] = pd.qcut(df['total night minutes'], num_intervals, labels=labels)
# Convertendo a coluna 'total night minutes' para códigos numéricos
df['total night minutes'] = df['total night minutes'].astype('category').cat.codes

# Descobrindo os percentis para criar os intervalos
labels_intl = ['intervalo1', 'intervalo2', 'intervalo3', 'intervalo4']
p25 = df['total intl minutes'].quantile(0.25)
p50 = df['total intl minutes'].quantile(0.5)
p75 = df['total intl minutes'].quantile(0.75)
# Definindo os intervalos usando os percentis
eve_minutes_bins = [df['total intl minutes'].min(), p25, p50, p75, df['total intl minutes'].max()]
# Realizando o binning usando o método pd.cut() diretamente na coluna 'total eve minutes'
df['total intl minutes'] = pd.cut(df['total intl minutes'], bins=eve_minutes_bins, labels=labels_intl)
# Convertendo a coluna 'total eve minutes' para códigos numéricos
df['total intl minutes'] = df['total intl minutes'].astype('category').cat.codes

In [254]:
# Dividindo os dados em conjuntos de treinamento e teste novamente após as transformações
X = df.drop(['state', 'area code', 'phone number', 'churn'], axis=1)

# Calculando a acurácia após as transformações
new_accuracy_after = train_and_evaluate_scaled(X, y)
print(f"Acurácia após transformações: {accuracy_after}")

# Comparando os resultados
print(f"Diferença na acurácia: {new_accuracy_after - accuracy_after}")

Acurácia após transformações: 0.9550224887556222
Diferença na acurácia: -0.011994002998500841


# Fazendo o mesmo processo para o arquivo “heart.csv”.

## 1. Ler os dados do arquivo “heart.csv” e fazer uma análise exploratória inicial, para conhecer os atributos

In [255]:
df2 = pd.read_csv("heart.csv")
df2.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [256]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [257]:
df2.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [258]:
df2.value_counts()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  ca  thal  target
38   1    2   138       175   0    1        173      0      0.0      2      4   2     1         2
59   1    0   110       239   0    0        142      1      1.2      1      1   3     0         1
          2   126       218   1    1        134      0      2.2      1      1   1     0         1
          1   140       221   0    1        164      1      0.0      2      0   2     1         1
          0   170       326   0    0        140      1      3.4      0      0   3     0         1
                                                                                               ..
51   1    2   94        227   0    1        154      1      0.0      2      1   3     1         1
          0   140       299   0    1        173      1      1.6      2      0   3     0         1
                        298   0    1        122      1      4.2      1      3   3     0         1
                        2

In [259]:
df2.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [260]:
for column in df2.columns:
    unique_values = df2[column].unique()
    print(f"Valores únicos na coluna '{column}': {unique_values}")

Valores únicos na coluna 'age': [63 37 41 56 57 44 52 54 48 49 64 58 50 66 43 69 59 42 61 40 71 51 65 53
 46 45 39 47 62 34 35 29 55 60 67 68 74 76 70 38 77]
Valores únicos na coluna 'sex': [1 0]
Valores únicos na coluna 'cp': [3 2 1 0]
Valores únicos na coluna 'trestbps': [145 130 120 140 172 150 110 135 160 105 125 142 155 104 138 128 108 134
 122 115 118 100 124  94 112 102 152 101 132 148 178 129 180 136 126 106
 156 170 146 117 200 165 174 192 144 123 154 114 164]
Valores únicos na coluna 'chol': [233 250 204 236 354 192 294 263 199 168 239 275 266 211 283 219 340 226
 247 234 243 302 212 175 417 197 198 177 273 213 304 232 269 360 308 245
 208 264 321 325 235 257 216 256 231 141 252 201 222 260 182 303 265 309
 186 203 183 220 209 258 227 261 221 205 240 318 298 564 277 214 248 255
 207 223 288 160 394 315 246 244 270 195 196 254 126 313 262 215 193 271
 268 267 210 295 306 178 242 180 228 149 278 253 342 157 286 229 284 224
 206 167 230 335 276 353 225 330 290 172 305 188 282 18

In [261]:
# Contando a quantidade de ocorrências de cada valor na coluna 'thal'
thal_counts = df2['thal'].value_counts()

# Exibindo os resultados
print(thal_counts)

2    166
3    117
1     18
0      2
Name: thal, dtype: int64


In [262]:
# Contando a quantidade de ocorrências de cada valor na coluna 'ca'
thal_counts = df2['ca'].value_counts()

# Exibindo os resultados
print(thal_counts)

0    175
1     65
2     38
3     20
4      5
Name: ca, dtype: int64


In [263]:
# Removendo linhas com valores de 'ca' fora do intervalo [0, 3] e valores de 'thal' iguais a 0
df2 = df2.query('0 <= ca <= 3 and thal != 0')

# Verificando o DataFrame após a remoção das linhas
for column in df2.columns:
    unique_values = df2[column].unique()
    print(f"Valores únicos na coluna '{column}': {unique_values}")

Valores únicos na coluna 'age': [63 37 41 56 57 44 52 54 48 49 64 58 50 66 43 69 59 42 61 40 71 51 65 53
 46 45 39 47 62 34 35 29 55 60 67 68 74 76 70 77 38]
Valores únicos na coluna 'sex': [1 0]
Valores únicos na coluna 'cp': [3 2 1 0]
Valores únicos na coluna 'trestbps': [145 130 120 140 172 150 110 135 160 105 125 142 155 104 138 108 134 122
 115 118 128 100 124  94 112 102 152 101 132 148 178 129 180 136 126 106
 156 170 146 117 200 165 174 192 144 123 154 114 164]
Valores únicos na coluna 'chol': [233 250 204 236 354 192 294 263 199 168 239 275 266 211 283 219 340 226
 247 234 243 302 212 175 417 197 198 177 273 213 304 232 269 360 308 245
 208 264 321 325 235 257 256 231 141 252 201 222 260 182 303 265 309 186
 203 183 220 209 258 227 261 221 205 240 318 298 564 277 214 248 255 207
 288 160 394 315 246 244 270 195 196 254 126 313 262 215 193 271 268 267
 210 295 306 178 223 242 180 228 149 278 253 342 157 286 229 284 224 206
 167 230 335 276 353 225 330 290 172 305 216 188 282 18

Dados relacionados a um estudo sobre doenças cardíacas. Cada amostra possui informações sobre um indivíduo e suas características médicas.

* **age**: Atributo numérico que representa a idade do paciente em anos.
* **sex**: Atributo categórico binário que indica o sexo do paciente (0 ou 1).
* **cp**: Atributo categórico que indica o tipo de dor no peito (valores de 0 a 3).
* **trestbps**: Atributo numérico que indica a pressão arterial em repouso do paciente em mm Hg.
* **chol**: Atributo numérico que representa o nível de colesterol sérico do paciente em mg/dl.
* **fbs**: Atributo categórico binário que indica se o paciente tem açúcar no sangue em jejum maior que 120 mg/dl (1) ou não (0).
* **restecg**: Atributo categórico que indica os resultados eletrocardiográficos em repouso (valores de 0 a 2).
* **thalach**: Atributo numérico que representa a frequência cardíaca máxima alcançada pelo paciente em bpm.
* **exang**: Atributo categórico binário que indica se o paciente teve angina induzida por exercício (1) ou não (0).
* **oldpeak**: Atributo numérico que indica a depressão do segmento ST induzida pelo exercício em relação ao repouso.
* **slope**: Atributo categórico que indica a inclinação do segmento ST no pico do exercício (valores de 0 a 2).
* **ca**: Atributo numérico que indica o número de vasos sanguíneos principais coloridos por fluoroscopia (valores de 0 a 3).
* **thal**: Atributo categórico que indica a presença de um defeito cardíaco reversível - talassemia (valores de 1 a 3).
* **target**: Atributo categórico binário que indica se o paciente foi diagnosticado com doença cardíaca (1) ou não (0).

## 2. Aplicar um algoritmo de classificação de sua escolha e calcular a precisão. A variável alvo para ser classificada é “target”

A variável "target" foi escolhida como variável alvo porque ela representa a classificação que queremos prever com base nos atributos disponíveis no conjunto de dados. Neste caso específico, "target" indica a presença (1) ou ausência (0) de uma doença cardíaca em um paciente.

Algoritmo de classificação escolhido: Random Forest

In [264]:
# Separando atributos e rótulo
X = df2.drop('target', axis=1)  
y = df2['target']  # variável alvo

In [265]:
accuracy_before = train_and_evaluate_scaled(X, y)
print(f"Acurácia antes das transformações: {accuracy_before:.4f}")

Acurácia antes das transformações: 0.8500


## 3. Realizar transformações (arredondamento, tornar binário, binning) em atributos numéricos que aparentemente podem se beneficiar dessas transformações

In [266]:
# 1 indica um oldpeak maior ou igual a 1 e 0 indica um oldpeak menor que 1.
df2['oldpeak'] = (df2['oldpeak'] >= 1).astype(int)

# Dividindo a coluna 'age' (idade) em intervalos (bins) baseados em quantis. 
df2['age'] = pd.qcut(df2['age'], 4, labels=False)

X = df2.drop('target', axis=1) # Atributos transformados

## 4. Aplicar novamente o algoritmo de classificação e comparar os resultados

In [267]:
# Realizando a divisão dos dados novamente para avaliar o impacto das transformações nos atributos numéricos no desempenho do modelo
# Aplicando o Random Forest novamente
# Calculando a acurácia após as transformações

accuracy_after = train_and_evaluate_scaled(X, y)
print(f"Acurácia após transformação: {accuracy_after:.4f}")

# Comparando os resultados
print(f"Diferença na acurácia: {accuracy_after - accuracy_before}")

Acurácia após transformação: 0.8500
Diferença na acurácia: 0.0


## 5. Repetir os passos 3 e 4 para diferentes atributos e formas de transformação

In [268]:
# Normalizando a coluna 'thalach' (frequência cardíaca máxima alcançada) usando a técnica Min-Max para que seus valores estejam no intervalo [0, 1].
min_thalach = df2['thalach'].min()
max_thalach = df2['thalach'].max()
df2['thalach'] = (df2['thalach'] - min_thalach) / (max_thalach - min_thalach)

# Transformação logarítmica na coluna 'chol' para reduzir o impacto de valores extremos de colesterol
df2['chol'] = np.log(df2['chol'])

# Transformando a coluna 'trestbps' para um valor binário, onde 1 indica uma pressão arterial igual ou maior que 140, e 0 indica uma pressão arterial menor que 140
df2['trestbps'] = (df2['trestbps'] >= 140).astype(int)

X = df2.drop('target', axis=1)

In [269]:
# Calculando a acurácia após as transformações
new_accuracy_after = train_and_evaluate_scaled(X, y)
print(f"Acurácia após transformação: {new_accuracy_after:.4f}")

# Comparando os resultados
print(f"Diferença na acurácia: {new_accuracy_after - accuracy_after}")

Acurácia após transformação: 0.8833
Diferença na acurácia: 0.033333333333333326
